rust/xaynet-sdk/src/client.rs

```diff
@@ -0,0 +1,194 @@
+use async_trait::async_trait;
+use std::rc::Rc;
+use thiserror::Error;
+use url::Url;
+
+use xaynet_core::{
+    common::RoundParameters,
+    crypto::{ByteObject, PublicSigningKey},
+    mask::Model,
+    SumDict,
+    UpdateSeedDict,
+};
+
+use crate::XaynetClient;
+
+#[derive(Debug, Error)]
+pub enum ClientError {
+    #[error("failed to deserialize data: {0}")]
+    Deserialize(String),
+
+    #[error("HTTP request failed: {0}")]
+    Http(String),
+
+    #[error("Unexpected response from the coordinator: expected {}, got {}", .0, .1)]
+    UnexpectedResponse(u32, u32),
+
+    #[error("Reading from file failed: {0}")]
+    Io(#[from] std::io::Error),
+
+    #[error("Unexpected certificate extension")]
+    UnexpectedCertificate,
+
+    #[error("No certificate found")]
+    NoCertificate,
+}
+
+impl ClientError {
+    fn http_error<E: std::error::Error>(e: E) -> Self {
+        Self::Http(format!("{}", e))
+    }
+}
+
+impl From<bincode::Error> for ClientError {
+    fn from(e: bincode::Error) -> Self {
+        Self::Deserialize(format!("{:?}", e))
+    }
+}
+
+impl From<::std::num::ParseIntError> for ClientError {
+    fn from(e: ::std::num::ParseIntError) -> Self {
+        Self::Deserialize(format!("{:?}", e))
+    }
+}
+
+/// A trait that Xaynet clients must implement
+#[async_trait]
+pub trait XaynetHttpClient {
+    type Error: std::error::Error;
+
+    async fn get(&mut self, url: &str) -> Result<Rc<[u8]>, ClientError>;
```

Comment:
The reason for this `Rc` was that I didn't to impose `Vec<u8>`, but I couldn't just do a `&[u8]`. But I think I found a better way to do this by using an associated `GetResponse` type actually.